### PR TITLE
A beginner's guide to installing and setting up the latest version of Nginx

### PR DESCRIPTION
**Pull request by EzequielBruni**
## Author checklist (to be completed by original Author)
- [x] Is this document a good fit for the Rocky project ?
- [ ] Is this a non-English contribution? 
- [x] Title and Author MetaTags have been inserted into the document 
- [x] If applicable, steps and instructions have been tested to work on a real system
- [x] Did you perform an initial self-review to fix basic typos and grammatical correctness

## Rocky Documentation checklist  (to be completed by Rocky team) 
- [x] 1st Pass (Check that document is good fit for project and author checklist completed)
- [x] 2nd Pass (Technical Review - check for technical correctness) 
- [x] 3rd Pass (Basic Editorial Review)
- [x] 4th Pass (Detailed Editorial Review and Peer Review)
- [ ] 5th Pass (Include document in TOC)
- [ ] Final pass/approval (Final Review)

### CODE REVIEW

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:126]: Why not using the service name? `firewall-cmd --add-service=http`
It's easier to understant.
```suggestion
sudo firewall-cmd --permanent --zone=public --add-service=http
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:138]:
```suggestion
sudo firewall-cmd --permanent --zone=public --add-service=https
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:147]: No it should not fail...
Maybe this is a remarque, so we could add a Note admonition

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:151]:
```suggestion
!!! Note

    Now, there’s a very small chance that this won’t work. In those rare cases, make `firewalld` do your bidding with the old turn-it-off-and-turn-it-on-again.

    ```bash
    systemctl restart firewalld
    ```
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:162]: If you choose to keep the service:

```suggestion
  services: cockpit dhcpv6-client ssh http https
  ports: 81/tcp 444/tcp 15151/tcp
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:179]:
```suggestion
You can also test that your web page is accessible with the command line :

```
curl -I http://myip/
```

```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:186]: Not at all!
And by the way selinux will prevent you from doing that.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:199]: This user should be a system user and not having a bash (/bin/nologin for example).

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:208]: Ah ah ! Good to see that sort of link!
Thx

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:231]: Great!

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:247]: Maybe an admonition?

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:268]: No! Never do that! My god!
You set as executable all your files! It's dangerous!

You should do something like that:
```
find /home/www -type d -exec chmod 755 "{}" \;
```
It that case, only directory are modified

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:272]: For a better security, only files or directory that you want to be writeable should belongs to the dedicated user. For the rest of the files, readonly (o=r) is sufficient to be served by the webserver.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:37]:
```suggestion
* A text editor of your choice, whether graphical or command-line based. For this tutorial, I’m using `nano`.
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:41]:
```suggestion
This part isn’t quite as simple as installing an extra repository usually is. We’re going to have to create a custom repo file for `dnf` to use, and download *Nginx* from. Technically, we’re sort of repurposing  repositories for CentOS that were made and hosted by *Nginx* themselves. This solution may or may not be viable in the long term, but it’s working great for now.
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:49]:
```suggestion
Now, make sure the `dnf-utils` package is installed, and any command-line text editor you may want:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:55]:
```suggestion
Once that’s all installed, power up your text editor of choice. You’ll want to create a file called (for the sake of simplicity) `nginx.repo`, and put it in `/etc/yum.repos.d/`. You can do this real quick like so:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:83]:
```suggestion
Save the file with <kbd>Control</kbd>-<kbd>S</kbd> (if using `nano`) and exit with <kbd>Control</kbd>-<kbd>X</kbd>.
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:93]:
```suggestion
Then, install the package `nginx` from the previously added repository:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:99]:
```suggestion
The terminal will ask you if you’re fine with installing the repository’s GPG key. You need that, so choose `Y` for yes.
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:101]:
```suggestion
Once the installation is done, start the `nginx` service and enable it to automatically start on reboot all in one go with:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:123]:
```suggestion
Here’s how to open up the necessary ports to actually see your web pages with `firewalld`, Rocky Linux’s default firewall with the `firewall-cmd` command. To add a new port, just run this:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:141]:
```suggestion
These configurations won’t take effect until you force the issue. To do that, tell `firewalld` to relead its configurations, like so:
```

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [inline docs/guides/web/nginx-mainline.md:210]: Don't give a password to a system user if you can.

**Review by alemorvan — CHANGES_REQUESTED** [review]
It's a good job @EzequielBruni 
But in my opinion, there are security aspects that need to be reworked.
I took the liberty of adding a few small comments and putting some commands between quotes.
Feel free to accept or refuse my suggestions!

**Comment by sspencerwire** [comment]
@EzequielBruni before I do any merging, etc., I'll let you and @alemorvan decide how to deal with @alemorvan suggestions. I did see one typo in my cursory glance through the text and conversation. It's in the conclusion section and I'm sure that you would have caught it at some point and wanted to fix it: 

## Conclusion

The basic installation and configuration of *Nginx* are easy

If they are installations and configurations, are is correct, but in this case, it should be "is"... yes??

This is going to be a great addition to the documentation. Thanks for doing this Ezequiel!!

**Comment by alemorvan** [comment]
As @EzequielBruni said at the end, a guide for multi-host miss.
I don't know well nginx, but I can help.

File: docs/guides/web/nginx-mainline.md
Comment (sspencerwire) [reply docs/guides/web/nginx-mainline.md:41]: I was just thinking about this: There is a module for nginx on Rocky for up-to version 1.20. Would this be better than a separate repository?

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:126]: For the service suggestion, I think I'll mention both. To be fair to me, I lifted this section from my upcoming guide on LXD web servers, and I'm pretty sure the ports matter there.

**Comment by EzequielBruni** [comment]
@alemorvan I was thinking that multi-site should be a separate document, and I can probably handle that, though of course I welcome your input. It'll need to cover making/organizing multiple config files, how to make a sites-enabled workflow, stuff like that. All-in-one documents have their advantages, but sometimes people see the length of a doc and think, "Man, I do NOT have time for this." 

Since no one's paying me by the article here, I thought it might be nice to break things down into more manageable steps.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:188]: selinux would prevent which part of those three paragraphs, specifically? I'm running my tests in containers that don't have it.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:41]: I know I mentioned this to you on Mattermost @sspencerwire, but here's the reference for posterity. The Nginx module is the current "stable" branch, and it's fine, but the makers of Nginx say this (in an admittedly older blog post):

"In NGINX nomenclature, 'stable' means that no new features are added (the feature set is stable). Only major bug fixes are committed to that version.

"We develop new features and all bug fixes in the mainline branch. We operate a time‑based release process, so you can expect to see new mainline releases approximately once per month, with exceptional releases when necessary. Over the last year, the mainline has seen the introduction of SPDY 3.1 support, authentication via subrequests, SSL session ticket support, IPv6 support for DNS, and PROXY protocol support. We integrated community contributions for SSL support for uwsgi. We’ve extended error logging, added cache revalidation directives, added SMTP pipelining, added buffering options for FastCGI, improved support for MP4 streaming, and extended handling of byte‑range requests for streaming and caching.

"Note that stable does not mean more reliable or more bug‑free. In fact, the mainline is generally regarded as more reliable because we port all bug fixes to it, and not just critical fixes as for the stable branch. On the other hand, changes in the stable branch are very unlikely to affect third‑party modules. We don’t make the same commitment concerning the mainline, where new features can affect the operation of third‑party modules."

And:

"We recommend that in general you deploy the NGINX mainline branch at all times. The main reason to use the stable branch is that you are concerned about possible impacts of new features, such as incompatibility with third‑party modules or the inadvertent introduction of bugs in new features."

- https://www.nginx.com/blog/nginx-1-6-1-7-released/

I'm thinking that I'll explain this further in the guide, and show how to install from the module as well for people.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:201]: Hmmm. I was taking use cases like client hosting accounts into consideration. Also, is there a specific change to the command to make it a system user?

Maybe we can talk about both use cases.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:212]: See above.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:270]: Hmm. I know I only mentioned static sites, but I was also thinking about stuff like PHP apps... but even those usually only require the files in certain directories to be executable. I'll go into a little more detail and include your suggestion.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:274]: Fair enough. Time to explain more again, I guess.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [reply docs/guides/web/nginx-mainline.md:126]: I think this is a rule that should be set at the editorial charter level. I propose that when the service is standard, we should use the name of the service, when the service is custom, we should use the port.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [reply docs/guides/web/nginx-mainline.md:188]: Is selinux in enforcing mode in your containers ?
SELinux will prevent any docroot that has not httpd_sys_content label (from my memory), and that not the case for /home. You will have to enable a special boolean to allow the docroot in /home.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [reply docs/guides/web/nginx-mainline.md:201]: I think most of the time, ftp or sftp user to access the docroot are virtual (or should be).
We can keep the id of a user creation, even not a system one, but the discourse must be clear on security. It's a good way to do it, but it's not the most secure way.

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [reply docs/guides/web/nginx-mainline.md:270]: No no! PHP files don't need to be executable at all. It's a common mistake. Just to be readable is enough.
No executable files inside a docroot please 🙏.
My rules on a production webserver (for example):
chown -R www:www /var/www/html
find /var/www/html -type d -exec chmod 555 "{}" \;
find /var/www/html -type f -exec chmod 444 "{}" \;
chmod 750 /var/www/html/cache/
chmod 640 /var/www/html/cache/*
(if no directory in cache of course or you will need to use find)

File: docs/guides/web/nginx-mainline.md
Comment (alemorvan) [reply docs/guides/web/nginx-mainline.md:41]: For my part, I am still not familiar with the module system to be activated in dnf. But this is the official way.
In this case, I think there can be the "official" version and then an explanation "If you want to have an even newer version"... and keep the external repository and the actual explanation which is fine.
But the officially preferred version by redhat first.
That's good, it adds information to the page.

Translated with www.DeepL.com/Translator (free version)

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:188]: selinux isn't installed at all in the containers. In fact, a lot of the guides I've tried have suggested /home/. If it's different for Rocky Linux (some of those guides *were* for Alpine), I can change it to somewhere in /usr/share maybe?

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:270]: Weird. Some apps I've used won't work at all without more permissive rules. but we can recommend what you say as the default. The help files for individual PHP apps can say differently if necessary.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:201]: Okay, all good stuff to know.

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:41]: Well there's Red Hat official, and there's "getting all the latest security patches" official. I lean toward the latter, and I think I'll be saying as much in the guide.

Comment (sspencerwire) [reply docs/guides/web/nginx-mainline.md:41]: @EzequielBruni I'm fine with whatever you feel here. Rocky being a bug-for-bug compatible version of Red Hat means that the code is always going to lag a bit. It's by design. Doing things with your repository is absolutely fine. We do this all the time when we add in third-party repositories to get newer versions of software. Best example here is PHP 8.x. So you can add in the module install as an option, or not. Totally up to you.

File: docs/guides/web/nginx-mainline.md
Comment (sspencerwire) [reply docs/guides/web/nginx-mainline.md:126]: Neither of you are wrong here. They are both (port or service) fundamentally the same. I want someone to show me how I can fine-tune firewalls down to a specific IP address allow rule. If they can show me this, and it's not a complete cluster, then I'll actually move away from iptables to firewalld.

File: docs/guides/web/nginx-mainline.md
Comment (sspencerwire) [reply docs/guides/web/nginx-mainline.md:188]: I understand what @EzequielBruni is doing here. The minimal container installs do not have SELinux. I'm sure it can be installed, but the whole point of those container installs is that they are by design, secured by the host OS.

File: docs/guides/web/nginx-mainline.md
Comment (sspencerwire) [reply docs/guides/web/nginx-mainline.md:201]: @EzequielBruni if you read over the vsftp guide that I wrote, this uses virtual users. Pretty cool!

File: docs/guides/web/nginx-mainline.md
Comment (EzequielBruni) [reply docs/guides/web/nginx-mainline.md:41]: Oh I think it's genuinely a good idea to have both methods outlined for the sake of clarity and education. I like being thorough where I can.

And to think this all started because I realized we would need a full Nginx install guide to refer to for the LXD guide I'm writing.

Kind of curious to see what Antoine will have to say about that one. :)

**Pull request merged**
